keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -292,18 +293,50 @@ is scoped to.
-  - This will only be available on mergable collection types such as objects (unless
-    `x-kubernetes-map-type=atomic`), maps with `x-kubernetes-map-type=granular` and lists
-    with `x-kubernetes-list-type` set to `set` or `map`.  See [Merge
-    Strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) for
-    details.
-  - The use of "old" is congruent with how `AdmissionReview` identifies the existing object as
-    `oldObject`.
-  - xref [analysis of possible interactions with immutability and
-    validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
+
+- For certain use cases (e.g. immmutability), rules may access to the existing version of the object (see [use cases and other considerations](#access-to-existing-object-state)).
+
+  - Not all update operations are considered comparable for the purposes of this mechanism:
+
+    - Both an old and a new value must exist. It will remain possible to check if a value has been
+      added or removed by placing update rules on the parent node.
+
+    - The path to the schema node being validated must unambiguously refer to the same node -- in
+      terms of identity -- in the new object as it did in the old object. For example, array items
+      and their descendants (spec.foo[10].bar), or maps marked
+      [atomic](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy).
+
+    - Semantics from server-side apply will be enforced during CRD writes, and errors will be
+      generated if an update rule is added to a schema node that is not comparable, e.g. "<path>:
+      update rule <rule> cannot be set on schema because the schema or its parent schema is not
+      mergeable". Updates to descendants of mergeable collection types (objects not marked
+      `x-kubernetes-map-type=atomic`, maps marked `x-kubernetes-map-type=granular` and lists with
+      `x-kubernetes-list-type` set to `set` or `map`) are candidates for validation using update
+      rules.
+
+  - An optional field on rule items, "updateRule," designates a CEL expression to be evaluated
+    instead of "rule" for comparable update operations. In addition to "self", the "updateRule"
+    expression may reference the identifier "oldSelf", which is guaranteed to be bound to a non-null
+    value at evaluation time.
+
+   ```yaml
+   x-kubernetes-validations:
+     - rule: "self in ['low', 'medium', 'high']"
+       updateRule: "self in ['low', 'medium', 'high'] && !(self in {'low':['high'],'high':['low']}[oldSelf])"
+   ```
+
+  - The expression in the "rule" field is still required and will be evaluated for operations that
+    are not a comparable update operations (as described above). Users can set the "rule" field to
+    the expression "true" if the rule is intended to apply solely to comparable update operations.
```

Comment:
why should it be required? It's syntactical sugar after all, but do we have a technical reason?

Reply:
That's an attempt to make it explicit to users that exactly one of the expressions will always be evaluated, and to head off mistakes like writing this:

```yaml
x-kubernetes-validations:
- updateRule: "self <= oldSelf && self < 100"
```

where the intent was 'values should always be less than 100 and never decrease'. Under the current proposal, that would produce an error when the CRD is updated. If the other expression were assumed to be "true", it would accept any value for self during creates / updates without comparable old values.

Reply:
ok, so UX. Fair enough.